include/dba.php

```diff
@@ -302,18 +352,32 @@ public function q($sql, $onlyquery = false) {
 		}
 
 		$r = array();
-		if ($this->mysqli) {
-			if ($result->num_rows) {
-				while($x = $result->fetch_array(MYSQLI_ASSOC))
+		switch ($this->driver) {
+			case 'pdo':
+				while ($x = $result->fetch(PDO::FETCH_ASSOC)) {
 					$r[] = $x;
+				}
+				$result->closeCursor();
+				break;
+			case 'mysqli':
+				while ($x = $result->fetch_array(MYSQLI_ASSOC)) {
+					$r[] = $x;
+				}
 				$result->free_result();
-			}
-		} else {
-			if (mysql_num_rows($result)) {
-				while($x = mysql_fetch_array($result, MYSQL_ASSOC))
+				break;
+			case 'mysql':
+				while ($x = mysql_fetch_array($result, MYSQL_ASSOC)) {
 					$r[] = $x;
+				}
 				mysql_free_result($result);
-			}
+				break;
+		}
+
+		// PDO doesn't return "true" on successful operations - like mysqli does
+		// Emulate this behaviour by checking if the query returned data and had columns
+		// This should be reliable enough
+		if (($this->driver == 'pdo') AND (count($r) == 0) AND ($columns == 0)) {
+			return true;
```

Comment:
No. Zero columns are returned as well if the query is something like "update" or "insert" and it was successful.